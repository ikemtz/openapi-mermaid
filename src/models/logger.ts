export interface ILogger {
  log: (data: string) => void;
  error: (data: string) => void;
  warn: (data: string) => void;
}
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-empty-function */
export class MockConsoleLogger implements ILogger {
  public log = (data: string) => {};
  public error = (data: string) => {};
  public warn = (data: string) => {};
}
