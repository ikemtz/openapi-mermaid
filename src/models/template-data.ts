export interface ITemplateData {
  entities?: IEntity[];
  paths: string[];
}
export interface IEntity {
  isEnum?: boolean;
  enumValues: (string | { key?: number; name: string })[];
  name: string;
  importTypes: IImportType[];
  valueProperties: IValueProperty[];
  referenceProperties: IReferenceProperty[];
}
export interface IImportType {
  name: string;
}
export interface IValueProperty {
  name: string;
  type?: string;
  isArray: boolean;
  hasMultipleValidators: boolean;
  required: boolean;
  maxLength?: number;
  minLength?: number;
}

export interface IReferenceProperty {
  name: string;
  referenceTypeName: string;
  isArray: boolean;
  required: boolean;
}
