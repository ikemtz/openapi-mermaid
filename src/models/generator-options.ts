/* eslint-disable @typescript-eslint/no-unused-vars */
import { resolve } from 'path';
import { ILogger } from './logger';
import { IEntity, IReferenceProperty, IValueProperty } from './template-data';

export interface IGeneratorOptions {
  logger?: ILogger;
  outputPath: string;
  outputFileName?: string;
  openApiJsonUrl?: string;
  openApiJsonFileName?: string;
  typeFilterCallBack?: (entity: IEntity, index: number, array: IEntity[]) => boolean;
  valuePropertyTypeFilterCallBack?: (valueProperty: IValueProperty, index: number, array: IValueProperty[]) => boolean;
  referencePropertyTypeFilterCallBack?: (referenceProperty: IReferenceProperty, index: number, array: IReferenceProperty[]) => boolean;
  templates?: ITemplates | null;
}
export interface ITemplates {
  htmlClassDiagram?: string;
  mdClassDiagram?: string;
}
export function defaultFilter(
  value: IEntity | IValueProperty | IReferenceProperty,
  index: number,
  array: IEntity[] | IValueProperty[] | IReferenceProperty[],
): boolean {
  return true;
}

export function setGeneratorOptionDefaults(options: IGeneratorOptions): IGeneratorOptions {
  const templateFolder = resolve(`${__dirname}/..`, 'templates');
  options.typeFilterCallBack = options.typeFilterCallBack ?? defaultFilter;
  options.valuePropertyTypeFilterCallBack = options.valuePropertyTypeFilterCallBack ?? defaultFilter;
  options.referencePropertyTypeFilterCallBack = options.referencePropertyTypeFilterCallBack ?? defaultFilter;
  options.templates = {
    ...options.templates,
    mdClassDiagram: options.templates?.mdClassDiagram ?? `${templateFolder}/class-diagram.md.hbs`,
    htmlClassDiagram: options.templates?.htmlClassDiagram ?? `${templateFolder}/class-diagram.html.hbs`,
  };
  return options;
}
