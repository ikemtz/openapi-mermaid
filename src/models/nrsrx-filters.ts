/* eslint-disable @typescript-eslint/no-unused-vars */
import { IEntity, IValueProperty } from './template-data';

export function nrsrxTypeFilterCallBack(val: IEntity, i: number, arr: IEntity[]): boolean {
  return !val.name.endsWith('ODataEnvelope');
}
export function nrsrxValuePropertyTypeFilterCallBack(val: IValueProperty, i: number, arr: IValueProperty[]): boolean {
  return !val.name.startsWith('created') && !val.name.startsWith('updated');
}
