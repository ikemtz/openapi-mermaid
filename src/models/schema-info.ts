import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { defaultFilter, IGeneratorOptions } from './generator-options';
import { IReferenceProperty, IValueProperty } from './template-data';

export class SchemaWrapperInfo {
  public propertySchemaObject: SchemaObject = {};
  public propertyReferenceObject: ReferenceObject = { $ref: '' };

  public readonly componentSchemaObject: SchemaObject;
  public readonly valueProperties: IValueProperty[];
  public referenceProperties: IReferenceProperty[];

  constructor(schemaItem: SchemaObject) {
    this.componentSchemaObject = schemaItem;
    this.valueProperties = [];
    this.referenceProperties = [];
  }

  public updateReferenceProperties(options: IGeneratorOptions): void {
    this.referenceProperties = this.referenceProperties.filter(options.referencePropertyTypeFilterCallBack || defaultFilter);
  }
}
