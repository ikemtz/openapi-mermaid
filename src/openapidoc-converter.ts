import { OpenAPIObject, ReferenceObject, SchemaObject } from 'openapi3-ts';
import { defaultFilter, IGeneratorOptions } from './models/generator-options';
import { SchemaWrapperInfo } from './models/schema-info';
import { IEntity, IImportType, IReferenceProperty, ITemplateData, IValueProperty } from './models/template-data';

export class OpenApiDocConverter {
  public readonly regex = /[A-z0-9]*$/s;
  constructor(private readonly options: IGeneratorOptions, private readonly apiDocument: OpenAPIObject) {}

  public convertDocument(): ITemplateData {
    const entities = this.convertEntities();
    const paths: string[] = [];
    return { entities, paths };
  }

  public convertEntities(): IEntity[] {
    const entities: IEntity[] = [];

    for (const schemaName in this.apiDocument.components?.schemas) {
      if (this.apiDocument.components?.schemas[schemaName]) {
        const schemaWrapperInfo = new SchemaWrapperInfo(this.apiDocument.components?.schemas[schemaName]);
        this.buildSchemaWrapperInfo(schemaWrapperInfo);
        schemaWrapperInfo.updateReferenceProperties(this.options);
        const entity = {
          name: schemaName,
          referenceProperties: schemaWrapperInfo.referenceProperties,
          valueProperties: schemaWrapperInfo.valueProperties.filter(this.options.valuePropertyTypeFilterCallBack || defaultFilter),
          importTypes: this.getImportTypes(schemaName, schemaWrapperInfo),
        };
        entities.push(entity);
      }
    }
    return entities.filter(this.options.typeFilterCallBack || defaultFilter);
  }

  public buildSchemaWrapperInfo(schemaWrapperInfo: SchemaWrapperInfo): void {
    for (const propertyName in schemaWrapperInfo.componentSchemaObject.properties) {
      if (
        (schemaWrapperInfo.propertySchemaObject = schemaWrapperInfo.componentSchemaObject.properties[propertyName] as SchemaObject).type && // NOSONAR
        schemaWrapperInfo.propertySchemaObject.type !== 'array'
      ) {
        schemaWrapperInfo.valueProperties.push(this.convertSchemaObjectToPropertyType(propertyName, schemaWrapperInfo));
      } else if (
        (schemaWrapperInfo.propertyReferenceObject = schemaWrapperInfo.componentSchemaObject.properties[propertyName] as ReferenceObject)
          .$ref
      ) {
        schemaWrapperInfo.referenceProperties.push(this.convertReferenceObjectToPropertyType(propertyName, schemaWrapperInfo));
      } else if (schemaWrapperInfo.propertySchemaObject.type === 'array' && schemaWrapperInfo.propertySchemaObject.items) {
        this.convertArray(propertyName, schemaWrapperInfo);
      }
    }
  }

  public convertArray(propertyName: string, schemaWrapperInfo: SchemaWrapperInfo): void {
    const arraySchemaObject = schemaWrapperInfo.propertySchemaObject.items as SchemaObject;
    if (arraySchemaObject.type) {
      schemaWrapperInfo.valueProperties.push(this.convertArrayObjectToValuePropertyType(propertyName, schemaWrapperInfo));
    } else {
      schemaWrapperInfo.propertyReferenceObject = schemaWrapperInfo.propertySchemaObject.items as ReferenceObject;
      schemaWrapperInfo.referenceProperties.push(this.convertArrayObjectToReferencePropertyType(propertyName, schemaWrapperInfo));
    }
  }

  public convertSchemaObjectToPropertyType(propertyName: string, schemaWrapperInfo: SchemaWrapperInfo): IValueProperty {
    const required = this.getIsRequired(propertyName, schemaWrapperInfo);
    return {
      required,
      name: propertyName,
      isArray: false,
      type: this.getPropertyType(schemaWrapperInfo),
      maxLength: schemaWrapperInfo.propertySchemaObject.maxLength,
      minLength: schemaWrapperInfo.propertySchemaObject.minLength,
      hasMultipleValidators:
        +required + +!!schemaWrapperInfo.propertySchemaObject.maxLength + +!!schemaWrapperInfo.propertySchemaObject.minLength > 1,
    };
  }

  public convertArrayObjectToValuePropertyType(propertyName: string, schemaWrapperInfo: SchemaWrapperInfo): IValueProperty {
    const required = this.getIsRequired(propertyName, schemaWrapperInfo);
    return {
      required,
      name: propertyName,
      isArray: true,
      type: this.getPropertyType(schemaWrapperInfo),
      hasMultipleValidators: false,
    };
  }

  public convertArrayObjectToReferencePropertyType(propertyName: string, schemaWrapperInfo: SchemaWrapperInfo): IReferenceProperty {
    return {
      ...this.convertReferenceObjectToPropertyType(propertyName, schemaWrapperInfo),
      isArray: true,
    };
  }

  public convertReferenceObjectToPropertyType(propertyName: string, schemaWrapperInfo: SchemaWrapperInfo): IReferenceProperty {
    return {
      name: propertyName,
      referenceTypeName: this.parseRef(schemaWrapperInfo),
      isArray: false,
      required: (schemaWrapperInfo.componentSchemaObject.required || []).indexOf(propertyName) > -1,
    };
  }

  public getPropertyType(schemaWrapperInfo: SchemaWrapperInfo): string {
    if (schemaWrapperInfo.propertySchemaObject.type === 'array' && schemaWrapperInfo.propertySchemaObject.items) {
      return (schemaWrapperInfo.propertySchemaObject.items as { type: string }).type;
    } else if (schemaWrapperInfo.propertySchemaObject.type === 'integer' && schemaWrapperInfo.propertySchemaObject.enum) {
      return 'string | number';
    } else if (schemaWrapperInfo.propertySchemaObject.type === 'integer') {
      return 'number';
    } else if (schemaWrapperInfo.propertySchemaObject.format === 'date-time') {
      return 'date';
    } else if (schemaWrapperInfo.propertySchemaObject.format === 'uuid') {
      return 'uuid';
    } else if (schemaWrapperInfo.propertySchemaObject.type === 'boolean') {
      return 'bool';
    }
    if (!schemaWrapperInfo.propertySchemaObject.type) {
      throw new Error('Invalid Propety Type');
    }
    return schemaWrapperInfo.propertySchemaObject.type;
  }

  public parseRef(schemaWrapperInfo: SchemaWrapperInfo): string {
    let regexResult: RegExpExecArray;
    let result: string | null = null;
    if (
      schemaWrapperInfo.propertyReferenceObject.$ref &&
      // tslint:disable-next-line: no-conditional-assignment
      (regexResult = this.regex.exec(schemaWrapperInfo.propertyReferenceObject.$ref) as RegExpExecArray) // NOSONAR
    ) {
      schemaWrapperInfo.propertyReferenceObject.$ref = regexResult[0];
      result = schemaWrapperInfo.propertyReferenceObject.$ref;
    }
    return result || 'unknown';
  }

  public getImportTypes(entityName: string, schemaWrapperInfo: SchemaWrapperInfo): IImportType[] {
    return schemaWrapperInfo.referenceProperties
      .map((t) => t.referenceTypeName)
      .filter((t) => t !== entityName)
      .filter((value, index, array) => array.indexOf(value) === index)
      .map((value) => ({ name: value }));
  }

  public getIsRequired(propertyName: string, schemaWrapperInfo: SchemaWrapperInfo): boolean {
    return (
      ((schemaWrapperInfo.componentSchemaObject.required || []).indexOf(propertyName) > -1 ||
        (schemaWrapperInfo.propertySchemaObject.nullable === undefined ? false : !schemaWrapperInfo.propertySchemaObject.nullable)) &&
      propertyName !== 'id'
    );
  }
}
