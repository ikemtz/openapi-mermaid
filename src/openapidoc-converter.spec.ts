import { OpenAPIObject } from 'openapi3-ts';
import { SchemaWrapperInfo } from './models/schema-info';
import { OpenApiDocConverter } from './openapidoc-converter';

describe('OpenApiDocConverter Tests', () => {
  it('should handle blank parseRef()', () => {
    const converter = new OpenApiDocConverter({ outputPath: '' }, {} as OpenAPIObject);
    const result = converter.parseRef({ propertyReferenceObject: {} } as SchemaWrapperInfo);
    expect(result).toBe('unknown');
  });

  it('should handle invalid getPropertyTypeScriptType()', (done) => {
    const converter = new OpenApiDocConverter({ outputPath: '' }, {} as OpenAPIObject);
    try {
      converter.getPropertyType({ propertySchemaObject: {} } as SchemaWrapperInfo);
      done.fail('This test should throw an exception');
    } catch (err) {
      expect(err).toMatchSnapshot();
      done();
    }
  });
});
