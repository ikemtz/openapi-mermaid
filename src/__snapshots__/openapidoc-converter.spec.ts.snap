// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenApiDocConverter Tests should handle invalid getPropertyTypeScriptType() 1`] = `[Error: Invalid Propety Type]`;
